cluster-autoscaler/cloudprovider/azure/azure_scale_set.go

```diff
@@ -253,27 +254,35 @@ func (scaleSet *ScaleSet) updateVMSSCapacity(future *azure.Future) {
 		return
 	}
 
-	klog.Errorf("virtualMachineScaleSetsClient.WaitForCreateOrUpdate for scale set %q failed: %v", scaleSet.Name, err)
+	klog.Errorf("virtualMachineScaleSetsClient.WaitForAsyncOperationResult for scale set %q failed: %v", scaleSet.Name, err)
 }
 
 // SetScaleSetSize sets ScaleSet size.
 func (scaleSet *ScaleSet) SetScaleSetSize(size int64) error {
 	scaleSet.sizeMutex.Lock()
 	defer scaleSet.sizeMutex.Unlock()
 
-	// Proactively set the VMSS size so autoscaler makes better decisions.
-	scaleSet.curSize = size
-	scaleSet.lastSizeRefresh = time.Now()
-
 	vmssInfo, rerr := scaleSet.getVMSSInfo()
 	if rerr != nil {
 		klog.Errorf("Failed to get information for VMSS (%q): %v", scaleSet.Name, rerr)
 		return rerr.Error()
 	}
 
+	// Abort scaling to avoid concurrent VMSS scaling if the VMSS is still under updating.
+	// Note that the VMSS provisioning state would be updated per scaleSet.sizeRefreshPeriod.
+	if vmssInfo.VirtualMachineScaleSetProperties != nil && strings.EqualFold(to.String(vmssInfo.VirtualMachineScaleSetProperties.ProvisioningState), string(compute.ProvisioningStateUpdating)) {
+		klog.Errorf("VMSS %q is still under updating, waiting for it finishes before scaling", scaleSet.Name)
+		return fmt.Errorf("VMSS %q is still under updating", scaleSet.Name)
+	}
+
+	// Proactively set the VMSS size so autoscaler makes better decisions.
```

Comment:
not a regression from this PR (master and 1.18) but we should move this to happen under `CreateOrUpdateAsync`, we shouldn't proactively update the size before sending the request.

What if `CreateOrUpdateAsync` fails because of a quota error, you'd have to wait for the next size refresh to clear.

Reply:
good catch, this should be be moved after `CreateOrUpdateAsync()`, so it fails then the cache still the same as before